src/twisted/test/test_defer.py

```diff
@@ -2446,3 +2446,301 @@ def test_cancelDeferUntilLockedWithTimeout(self):
         self.assertFalse(timeoutCall.active())
         self.assertIsNone(self.lock._timeoutCall)
         self.failureResultOf(deferred, defer.CancelledError)
+
+
+
+def _overrideFunc(v, t):
+    """
+    Private function to be used to pass as an alternate onTimeoutCancel value
+    to timeoutDeferred
+    """
+    return "OVERRIDDEN"
+
+
+
+class DeferredAddTimeoutTests(unittest.SynchronousTestCase):
+    """
+    Tests for the function L{Deferred.addTimeout}
+    """
+    def test_timeoutChainable(self):
+        """
+        L{defer.Deferred.addTimeout} returns its own L{defer.Deferred} so it
+        can be called in a callback chain.
+        """
+        d = defer.Deferred().addTimeout(5, Clock()).addCallback(lambda _: "done")
+        d.callback(None)
+        self.assertEqual("done", self.successResultOf(d))
+
+
+    def test_successResultBeforeTimeout(self):
+        """
+        The L{defer.Deferred} callbacks with the result if it succeeds before
+        the timeout. No cancellation happens after the callback either,
+        which could also cancel inner deferreds.
+        """
+        clock = Clock()
+        d = defer.Deferred()
+        d.addTimeout(10, clock)
+
+        # timeout happens before inner deferred is returned as a errback on d
```

Comment:
Ah sorry, I guess I meant to say "the timeout is added before...".  You're right, the timing out actually happens afterward.  Will fix, thanks!